README.md
<div align="center">
    <img src="https://user-images.githubusercontent.com/6267663/230347878-f2873a58-f578-4e95-86e0-7bebfd78f4f1.svg" width="300">
    <h1>Orders Typescript SDK</h1>
   <p>An effortless integration. Designed for growth.</p>
   <a href="https://docs.klarna.com/"><img src="https://img.shields.io/static/v1?label=Docs&message=API Ref&color=000&style=for-the-badge" /></a>
   <a href="https://github.com/speakeasy-sdks/klarna-ts/actions"><img src="https://img.shields.io/github/actions/workflow/status/speakeasy-sdks/klarna-ts/speakeasy_sdk_generation.yml?style=for-the-badge" /></a>
  <a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-blue.svg?style=for-the-badge" /></a>
  <a href="https://github.com/speakeasy-sdks/klarna-ts/releases"><img src="https://img.shields.io/github/v/release/speakeasy-sdks/klarna-ts?sort=semver&style=for-the-badge" /></a>
</div>

<!-- Start SDK Installation -->
## SDK Installation

### NPM

```bash
npm add @speakeasy-sdks/klarna-orders
```

### Yarn

```bash
yarn add @speakeasy-sdks/klarna-orders
```
<!-- End SDK Installation -->

## SDK Example Usage
<!-- Start SDK Example Usage -->
```typescript
import { Klarna } from "@speakeasy-sdks/klarna-orders";
import { CreateOrderResponse } from "@speakeasy-sdks/klarna-orders/dist/sdk/models/operations";
import {
  AuthorizedPaymentMethodTypeEnum,
  CustomerOrganizationEntityTypeEnum,
  SubscriptionIntervalEnum,
} from "@speakeasy-sdks/klarna-orders/dist/sdk/models/shared";

const sdk = new Klarna({
  security: {
    apiKeyAuth: "YOUR_BEARER_TOKEN_HERE",
  },
});

sdk.orders.read({
  authorizationToken: "corrupti",
  createOrderRequestInput: {
    autoCapture: false,
    billingAddress: {
      attention: "Attn",
      city: "London",
      country: "GB",
      email: "test.sam@test.com",
      familyName: "Andersson",
      givenName: "Adam",
      organizationName: "provident",
      phone: "+44795465131",
      postalCode: "W1G 0PW",
      region: "OH",
      streetAddress: "33 Cavendish Square",
      streetAddress2: "Floor 22 / Flat 2",
      title: "Mr.",
    },
    customPaymentMethodIds: [
      "quibusdam",
      "unde",
      "nulla",
    ],
    customer: {
      dateOfBirth: "1978-12-31",
      gender: "male",
      lastFourSsn: "corrupti",
      nationalIdentificationNumber: "illum",
      organizationEntityType: CustomerOrganizationEntityTypeEnum.LimitedPartnership,
      organizationRegistrationId: "error",
      title: "Mr.",
      type: "organization",
      vatId: "deserunt",
    },
    locale: "en-GB",
    merchantData: "{"order_specific":[{"substore":"Women's Fashion","product_name":"Women Sweatshirt"}]}",
    merchantReference1: "ON4711",
    merchantReference2: "hdt53h-zdgg6-hdaff2",
    merchantUrls: {
      authorization: "https://www.example-url.com/authorization",
      confirmation: "https://www.example-url.com/confirmation",
      notification: "https://www.example-url.com/notification",
      push: "https://www.example-url.com/push",
    },
    orderAmount: 2500,
    orderLines: [
      {
        imageUrl: "https://www.exampleobjects.com/logo.png",
        merchantData: "{"customer_account_info":[{"unique_account_identifier":"test@gmail.com","account_registration_date":"2017-02-13T10:49:20Z","account_last_modified":"2019-03-13T11:45:27Z"}]}",
        name: "Running shoe",
        productIdentifiers: {
          brand: "shoe-brand",
          categoryPath: "Shoes > Running",
          color: "white",
          globalTradeItemNumber: "4912345678904",
          manufacturerPartNumber: "AD6654412-334.22",
          size: "small",
        },
        productUrl: "https://.../AD6654412.html",
        quantity: 1,
        quantityUnit: "pcs",
        reference: "AD6654412",
        subscription: {
          interval: SubscriptionIntervalEnum.Week,
          intervalCount: 297534,
          name: "Larry Windler",
        },
        taxRate: 1900,
        totalAmount: 2500,
        totalDiscountAmount: 500,
        totalTaxAmount: 475,
        type: "physical",
        unitPrice: 2500,
      },
      {
        imageUrl: "https://www.exampleobjects.com/logo.png",
        merchantData: "{"customer_account_info":[{"unique_account_identifier":"test@gmail.com","account_registration_date":"2017-02-13T10:49:20Z","account_last_modified":"2019-03-13T11:45:27Z"}]}",
        name: "Running shoe",
        productIdentifiers: {
          brand: "shoe-brand",
          categoryPath: "Shoes > Running",
          color: "white",
          globalTradeItemNumber: "4912345678904",
          manufacturerPartNumber: "AD6654412-334.22",
          size: "small",
        },
        productUrl: "https://.../AD6654412.html",
        quantity: 1,
        quantityUnit: "pcs",
        reference: "AD6654412",
        subscription: {
          interval: SubscriptionIntervalEnum.Week,
          intervalCount: 791725,
          name: "Ken Kshlerin",
        },
        taxRate: 1900,
        totalAmount: 2500,
        totalDiscountAmount: 500,
        totalTaxAmount: 475,
        type: "physical",
        unitPrice: 2500,
      },
    ],
    orderTaxAmount: 475,
    purchaseCountry: "GB",
    purchaseCurrency: "GBP",
    shippingAddress: {
      attention: "Attn",
      city: "London",
      country: "GB",
      email: "test.sam@test.com",
      familyName: "Andersson",
      givenName: "Adam",
      organizationName: "recusandae",
      phone: "+44795465131",
      postalCode: "W1G 0PW",
      region: "OH",
      streetAddress: "33 Cavendish Square",
      streetAddress2: "Floor 22 / Flat 2",
      title: "Mr.",
    },
  },
}).then((res: CreateOrderResponse) => {
  if (res.statusCode == 200) {
    // handle response
  }
});
```
<!-- End SDK Example Usage -->

<!-- Start SDK Available Operations -->
## Available Resources and Operations


### [authorizations](docs/authorizations/README.md)

* [cancel](docs/authorizations/README.md#cancel) - Cancel an existing authorization

### [orders](docs/orders/README.md)

* [read](docs/orders/README.md#read) - Create a new order

### [sessions](docs/sessions/README.md)

* [read](docs/sessions/README.md#read) - Read an existing payment session
* [read](docs/sessions/README.md#read) - Create a new payment session
* [update](docs/sessions/README.md#update) - Update an existing payment session

### [tokens](docs/tokens/README.md)

* [purchase](docs/tokens/README.md#purchase) - Generate a consumer token
<!-- End SDK Available Operations -->

### Maturity

This SDK is in beta, and there may be breaking changes between versions without a major version update. Therefore, we recommend pinning usage
to a specific package version. This way, you can install the same version each time without breaking changes unless you are intentionally
looking for the latest version.

### Contributions

While we value open-source contributions to this SDK, this library is generated programmatically.
Feel free to open a PR or a Github issue as a proof of concept and we'll do our best to include it in a future release !

### SDK Created by [Speakeasy](https://docs.speakeasyapi.dev/docs/using-speakeasy/client-sdks)



RELEASES.md


## 2023-04-18 06:59:57
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.20.1 (2.18.1) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.0.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.0.0 - .

## 2023-04-18 07:06:15
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.20.1 (2.18.1) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.0.1] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.0.1 - .

## 2023-04-19 00:27:08
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.20.2 (2.18.2) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.0.2] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.0.2 - .

## 2023-04-21 00:25:36
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.21.4 (2.19.1) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.1.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.1.0 - .

## 2023-04-22 00:25:52
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.22.1 (2.20.1) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.2.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.2.0 - .

## 2023-04-26 00:26:00
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.23.1 (2.21.1) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.3.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.3.0 - .

## 2023-04-27 00:29:25
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.25.1 (2.22.0) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.4.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.4.0 - .

## 2023-04-28 00:26:02
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.26.2 (2.23.2) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.5.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.5.0 - .

## 2023-04-29 00:25:12
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.26.3 (2.23.3) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.5.1] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.5.1 - .

## 2023-04-30 00:28:08
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.26.4 (2.23.4) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.5.2] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.5.2 - .

## 2023-05-02 00:25:21
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.26.5 (2.23.6) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.5.3] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.5.3 - .

## 2023-05-03 00:24:59
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.27.0 (2.24.0) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.6.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.6.0 - .

## 2023-05-05 00:24:24
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.29.0 (2.26.0) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.7.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.7.0 - .

## 2023-05-06 00:23:41
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.29.1 (2.26.1) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.7.1] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.7.1 - .

## 2023-05-10 00:25:06
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.29.2 (2.26.2) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.7.2] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.7.2 - .

## 2023-05-11 00:25:09
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.30.0 (2.26.3) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.7.3] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.7.3 - .

## 2023-05-12 00:25:06
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.30.1 (2.26.4) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.7.4] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.7.4 - .

## 2023-05-13 00:23:37
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.31.1 (2.27.0) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.8.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.8.0 - .

## 2023-05-16 00:26:16
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.32.0 (2.28.0) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.9.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.9.0 - .

## 2023-05-17 00:29:05
### Changes
Based on:
- OpenAPI Doc 1.0.0 ./openapi/openapi.yaml
- Speakeasy CLI 1.33.2 (2.29.0) https://github.com/speakeasy-api/speakeasy
### Releases
- [NPM v2.10.0] https://www.npmjs.com/package/@speakeasy-sdks/klarna-orders/v/2.10.0 - .

src/sdk/models/shared/sessionread.ts
/*
 * Code generated by Speakeasy (https://speakeasyapi.dev). DO NOT EDIT.
 */

import { SpeakeasyBase, SpeakeasyMetadata } from "../../../internal/utils";
import { Address } from "./address";
import { Attachment } from "./attachment";
import { CustomerRead } from "./customerread";
import { MerchantUrls } from "./merchanturls";
import { Options } from "./options";
import { OrderLine } from "./orderline";
import { PaymentMethodCategory } from "./paymentmethodcategory";
import { Expose, Transform, Type } from "class-transformer";

/**
 * The acquiring channel in which the session takes place. Ecommerce is default unless specified. Any other values should be defined in the agreement.
 */
export enum SessionReadAcquiringChannelEnum {
  Ecommerce = "ECOMMERCE",
  InStore = "IN_STORE",
  Telesales = "TELESALES",
}

/**
 * Intent for the session. The field is designed to let partners inform Klarna of the purpose of the customer’s session.
 */
export enum SessionReadIntentEnum {
  Buy = "buy",
  Tokenize = "tokenize",
  BuyAndTokenize = "buy_and_tokenize",
}

/**
 * The current status of the session. Possible values: 'complete', 'incomplete' where 'complete' is set when the order has been placed.
 */
export enum SessionReadStatusEnum {
  Complete = "complete",
  Incomplete = "incomplete",
}

/**
 * successful operation
 */
export class SessionRead extends SpeakeasyBase {
  /**
   * The acquiring channel in which the session takes place. Ecommerce is default unless specified. Any other values should be defined in the agreement.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "acquiring_channel" })
  acquiringChannel?: SessionReadAcquiringChannelEnum;

  @SpeakeasyMetadata()
  @Expose({ name: "attachment" })
  @Type(() => Attachment)
  attachment?: Attachment;

  /**
   * Authorization token.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "authorization_token" })
  authorizationToken?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "billing_address" })
  @Type(() => Address)
  billingAddress?: Address;

  /**
   * Token to be passed to the JS client
   */
  @SpeakeasyMetadata()
  @Expose({ name: "client_token" })
  clientToken?: string;

  /**
   * Promo codes - The array could be used to define which of the configured payment options within a payment category (pay_later, pay_over_time, etc.) should be shown for this purchase. Discuss with the delivery manager to know about the promo codes that will be configured for your account. The feature could also be used to provide promotional offers to specific customers (eg: 0% financing). Please be informed that the usage of this feature can have commercial implications.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "custom_payment_method_ids" })
  customPaymentMethodIds?: string[];

  @SpeakeasyMetadata()
  @Expose({ name: "customer" })
  @Type(() => CustomerRead)
  customer?: CustomerRead;

  /**
   * Design package to use in the session. This can only by used if a custom design has been implemented for Klarna Payments and agreed upon in the agreement. It might have a financial impact. Delivery manager will provide the value for the parameter.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "design" })
  design?: string;

  /**
   * Session expiration date
   */
  @SpeakeasyMetadata()
  @Expose({ name: "expires_at" })
  @Transform(({ value }) => new Date(value), { toClassOnly: true })
  expiresAt?: Date;

  /**
   * Intent for the session. The field is designed to let partners inform Klarna of the purpose of the customer’s session.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "intent" })
  intent?: SessionReadIntentEnum;

  /**
   * Used to define the language and region of the customer. The locale follows the format of [RFC 1766](https://datatracker.ietf.org/doc/rfc1766/), meaning its value consists of language-country.
   *
   * @remarks
   * Read more on **[Supported Locals and Currencies](https://docs.klarna.com/klarna-payments/in-depth-knowledge/puchase-countries-currencies-locales/)**.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "locale" })
  locale?: string;

  /**
   * Pass through field to send any information about the order to be used later for reference while retrieving the order details (max 6000 characters)
   */
  @SpeakeasyMetadata()
  @Expose({ name: "merchant_data" })
  merchantData?: string;

  /**
   * Used for storing merchant's internal order number or other reference.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "merchant_reference1" })
  merchantReference1?: string;

  /**
   * Used for storing merchant's internal order number or other reference. The value is available in the settlement files. (max 255 characters).
   */
  @SpeakeasyMetadata()
  @Expose({ name: "merchant_reference2" })
  merchantReference2?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "merchant_urls" })
  @Type(() => MerchantUrls)
  merchantUrls?: MerchantUrls;

  @SpeakeasyMetadata()
  @Expose({ name: "options" })
  @Type(() => Options)
  options?: Options;

  /**
   * Total amount of the order including tax and any available discounts. The value should be in non-negative minor units. Eg: 25 Euros should be 2500.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "order_amount" })
  orderAmount?: number;

  /**
   * The array containing list of line items that are part of this order. Maximum of 1000 line items could be processed in a single order.
   */
  @SpeakeasyMetadata({ elemType: OrderLine })
  @Expose({ name: "order_lines" })
  @Type(() => OrderLine)
  orderLines?: OrderLine[];

  /**
   * Total tax amount of the order. The value should be in non-negative minor units. Eg: 25 Euros should be 2500.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "order_tax_amount" })
  orderTaxAmount?: number;

  /**
   * Available payment method categories
   */
  @SpeakeasyMetadata({ elemType: PaymentMethodCategory })
  @Expose({ name: "payment_method_categories" })
  @Type(() => PaymentMethodCategory)
  paymentMethodCategories?: PaymentMethodCategory[];

  /**
   * The purchase country of the customer. The billing country always overrides purchase country if the values are different. Formatted according to ISO 3166 alpha-2 standard, e.g. GB, SE, DE, US, etc.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "purchase_country" })
  purchaseCountry?: string;

  /**
   * The purchase currency of the order. Formatted according to ISO 4217 standard, e.g. USD, EUR, SEK, GBP, etc.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "purchase_currency" })
  purchaseCurrency?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "shipping_address" })
  @Type(() => Address)
  shippingAddress?: Address;

  /**
   * The current status of the session. Possible values: 'complete', 'incomplete' where 'complete' is set when the order has been placed.
   */
  @SpeakeasyMetadata()
  @Expose({ name: "status" })
  status?: SessionReadStatusEnum;
}
